akka-stream-tests/src/test/scala/akka/stream/scaladsl/UnfoldResourceAsyncSourceSpec.scala

```diff
@@ -0,0 +1,247 @@
+
+      expectMsg(Open)
+      createPromise.success(resource)
+
+      sub.cancel()
+      expectMsg(Close(resource))
+      resource.close()
+      closePromise.success(Done)
+    }
+
+    "continue when Strategy is Resume and exception happened" in assertAllStagesStopped {
+      val p = Source.unfoldResourceAsync[String, BufferedReader](open,
+        reader ⇒ {
+          val s = reader.readLine()
+          if (s != null && s.contains("b")) throw ex else Promise.successful(Option(s)).future
+        }, close).withAttributes(supervisionStrategy(resumingDecider))
+        .runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[String]()
+
+      p.subscribe(c)
+      val sub = c.expectSubscription()
+
+      (0 to 49).foreach(i ⇒ {
+        sub.request(1)
+        c.expectNext() should ===(if (i < 10) manyLinesArray(i) else manyLinesArray(i + 10))
+      })
+      sub.request(1)
+      c.expectComplete()
+    }
+
+    "close and open stream again when Strategy is Restart" in assertAllStagesStopped {
+      val p = Source.unfoldResourceAsync[String, BufferedReader](open,
+        reader ⇒ {
+          val s = reader.readLine()
+          if (s != null && s.contains("b")) throw ex else Promise.successful(Option(s)).future
+        }, close).withAttributes(supervisionStrategy(restartingDecider))
+        .runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[String]()
+
+      p.subscribe(c)
+      val sub = c.expectSubscription()
+
+      (0 to 19).foreach(i ⇒ {
+        sub.request(1)
+        c.expectNext() should ===(manyLinesArray(0))
+      })
+      sub.cancel()
+    }
+
+    "work with ByteString as well" in assertAllStagesStopped {
+      val chunkSize = 50
+      val buffer = Array.ofDim[Char](chunkSize)
+      val p = Source.unfoldResourceAsync[ByteString, Reader](open,
+        reader ⇒ {
+          val p = Promise[Option[ByteString]]
+          val s = reader.read(buffer)
+          if (s > 0) p.success(Some(ByteString(buffer.mkString("")).take(s))) else p.success(None)
+          p.future
+        },
+        reader ⇒ {
+          reader.close()
+          Promise.successful(Done).future
+        }).runWith(Sink.asPublisher(false))
+      val c = TestSubscriber.manualProbe[ByteString]()
+
+      var remaining = manyLines
+      def nextChunk() = {
+        val (chunk, rest) = remaining.splitAt(chunkSize)
+        remaining = rest
+        chunk
+      }
+
+      p.subscribe(c)
+      val sub = c.expectSubscription()
+
+      (0 to 121).foreach(i ⇒ {
+        sub.request(1)
+        c.expectNext().utf8String should ===(nextChunk().toString)
+      })
+      sub.request(1)
+      c.expectComplete()
+    }
+
+    "use dedicated blocking-io-dispatcher by default" in assertAllStagesStopped {
+      val sys = ActorSystem("dispatcher-testing", UnboundedMailboxConfig)
+      val materializer = ActorMaterializer()(sys)
+      implicit val timeout = Timeout(500.millis)
```

Comment:
where is this timeout used? 500 millis might be to short in some build environments
perhaps use `remainingOrDefault`?